velox/dwio/parquet/reader/ParquetReader.cpp

```diff
@@ -117,24 +117,23 @@ ParquetRowReader::ParquetRowReader(
   }
 
   if (options.getScanSpec()) {
-    auto& scanSpec = *options.getScanSpec();
-    for (auto& colSpec : scanSpec.children()) {
+    auto& childrenScanSpec = options.getScanSpec()->children();
+    for (uint64_t colIdx = 0; colIdx < childrenScanSpec.size(); colIdx++) {
+      auto& colSpec = childrenScanSpec[colIdx];
+
       VELOX_CHECK(
           !colSpec->extractValues(),
           "Subfield access is NYI in parquet reader");
       if (colSpec->filter()) {
-        // TODO: remove linear search
-        uint64_t colIdx = std::find(
-                              reader_->names.begin(),
-                              reader_->names.end(),
-                              colSpec->fieldName()) -
-            reader_->names.begin();
         VELOX_CHECK(
-            colIdx < reader_->names.size(),
+            columnIds[colIdx] < reader_->names.size(),
             "Unexpected columns name: {}",
             colSpec->fieldName());
         toDuckDbFilter(
-            colIdx, reader_->return_types[colIdx], colSpec->filter(), filters_);
+            colIdx,
+            reader_->return_types[columnIds[colIdx]],
```

Comment:
I don't believe there are any guarantees on the order of ScanSpecs in childrenScanSpec, hence, it is fragile to use an colIdx here as an index into columnIds. I expect you need to lookup by name.

Reply:
> I don't believe there are any guarantees on the order of ScanSpecs in childrenScanSpec, hence, it is fragile to use an colIdx here as an index into columnIds. I expect you need to lookup by name.

The children in ScanSpec was initially created from the output types of the TableScan operator. Then the filter columns, if not in the output column list, would be appended afterwards. This order is deterministic. Then the selector is created based on the ScanSpecs in HiveDataSource::addSplit(), with the same order as the ScanSpecs, minus the constant columns. DWRF reader could be reordering the ScanSpecs, but Parquet doesn't have reordering, and for every call to DuckDB reader, the order of ScanSpec is deterministic. 

But when there're partition key column in the output list, the ScanSpecs would contain the partition keys while the Selector doesn't. The columnIds should not contain the partition key as the DuckDB only creates ColumnReaders for the columns in the file. The filter index should be columnIds index, and you are right that we need to search the ScanSpecs to get the filter. So I changed the code to

```
for (uint64_t i = 0; i < filter.size(); i++) {
    uint64_t columnId = filter[i].column;
    VELOX_CHECK(
        columnId < reader_->names.size(),
        "Unexpected column Id {} for name: {}",
        columnId,
        filter[i].name);
    columnIds.push_back(columnId);

    // DuckDB ParquetReader return_types is actually the type for all columns in
    // the file
    ::duckdb::LogicalType duckDbType = reader_->return_types[columnId];
    duckdbRowType_.push_back(duckDbType);

    auto veloxFilter = getFilter(filter[i].name, options.getScanSpec());
    if (veloxFilter) {
      toDuckDbFilter(i, duckDbType, veloxFilter, filters_);
    }
  }

  std::vector<idx_t> groups;
  for (idx_t i = 0; i < reader_->NumRowGroups(); i++) {
    auto groupOffset = reader_->GetFileMetadata()->row_groups[i].file_offset;
    if (groupOffset >= options.getOffset() &&
        groupOffset < (options.getLength() + options.getOffset())) {
      groups.push_back(i);
    }
  }
```